### PR TITLE
feat(snowflake): implement and test

### PR DESCRIPTION
**Pull request by Favna**
closes #6 
closes #7 
closes #9 
closes #10

### CODE REVIEW

File: packages/snowflake/src/lib/DiscordSnowflake.ts [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+	 * @param id The id to parse
+	 */
+	public constructor(id: string | bigint) {
+		super(id, DiscordSnowflake.DiscordEpoch);
+	}
+
+	/**
+	 * Discord epoch (2015-01-01T00:00:00.000Z)
+	 */
+	public static DiscordEpoch = 1420070400000;
```
Comment (vladfrangu) [inline packages/snowflake/src/lib/DiscordSnowflake.ts:14]: readonly, and rename to just `Epoch`

**Review by kyranet — DISMISSED** [review]

File: packages/snowflake/src/lib/Snowflake.ts [hunk 2]
```diff
@@ -0,0 +1,86 @@
+let INCREMENT = 0n;
+
+/**
+ * A class for parsing snowflake ids
+ */
+export class Snowflake {
+	/**
+	 * The id in BigInt form
+	 */
+	public id: bigint;
```
Comment (vladfrangu) [inline packages/snowflake/src/lib/Snowflake.ts:10]: readonly all of them

File: packages/snowflake/src/lib/Snowflake.ts [hunk 3]
```diff
@@ -0,0 +1,86 @@
+let INCREMENT = 0n;
```
Comment (vladfrangu) [inline packages/snowflake/src/lib/Snowflake.ts:1]: Make this a hard private per class instance maybe? That way, different epochs won't use the same increment

File: packages/snowflake/src/lib/Snowflake.ts [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+
+	/**
+	 * The increment stored in the snowflake
+	 */
+	public increment: number;
+
+	/**
+	 * The epoch to use in the snowflake
+	 */
+	public epoch: number | bigint;
```
Comment (vladfrangu) [inline packages/snowflake/src/lib/Snowflake.ts:35]: make this always a bigint

File: packages/snowflake/src/lib/Snowflake.ts [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+	 * @param timestamp Timestamp or date of the snowflake to generate
+	 */
+	public static generate(epoch: number, timestamp: number | Date = Date.now()): Snowflake {
+		if (timestamp instanceof Date) timestamp = timestamp.getTime();
+		if (typeof timestamp !== 'number' || Number.isNaN(timestamp)) {
+			throw new TypeError(`"timestamp" argument must be a number or Date (received ${Number.isNaN(timestamp) ? 'NaN' : typeof timestamp})`);
+		}
+		if (INCREMENT >= 4095n) INCREMENT = 0n;
+
+		// timestamp, workerID hard-coded to 1, processID hard-coded to 1, increment
```
Comment (vladfrangu) [inline packages/snowflake/src/lib/Snowflake.ts:83]: Allow us to change both of these as an option please, enforcing bigints

File: packages/snowflake/src/lib/TwitterSnowflake.ts [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+	 */
+	public constructor(id: string | bigint) {
+		super(id, TwitterSnowflake.TwitterEpoch);
+	}
+
+	/**
+	 * Twitter epoch (2006-03-21T20:50:14.000Z)
+	 * @see https://twitter.com/jack/status/20, first tweet ever made
+	 */
+	public static TwitterEpoch = 1142974214000;
```
Comment (vladfrangu) [inline packages/snowflake/src/lib/TwitterSnowflake.ts:15]: Ditto as for Discord

File: packages/snowflake/src/lib/Snowflake.ts [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+	/**
+	 * The epoch to use in the snowflake
+	 */
+	public epoch: number | bigint;
+
+	/**
+	 * @param id The id to parse
+	 * @param epoch the epoch to use
+	 */
+	public constructor(id: string | bigint, epoch: number) {
```
Comment (vladfrangu) [inline packages/snowflake/src/lib/Snowflake.ts:41]: Epoch should be before the ID

**Review by vladfrangu — CHANGES_REQUESTED** [review]

File: packages/snowflake/src/lib/Snowflake.ts (hunk 3 above)
Comment (kyranet) [reply packages/snowflake/src/lib/Snowflake.ts:1]: ... hard private static...?

Comment (vladfrangu) [reply packages/snowflake/src/lib/Snowflake.ts:1]: No. just a `#increment`

File: packages/snowflake/src/lib/Snowflake.ts (hunk 4 above)
Comment (Favna) [reply packages/snowflake/src/lib/Snowflake.ts:35]: image

File: packages/snowflake/src/lib/Snowflake.ts (hunk 3 above)
Comment (vladfrangu) [reply packages/snowflake/src/lib/Snowflake.ts:1]: Ohhh, wait, duhh, I forgot to say this (welcome to dumb brain)

My idea was you create a Snowflake class with the epoch ONCE, then run, say .deconstruct / .generate on the instance, so no statics

Comment (Favna) [reply packages/snowflake/src/lib/Snowflake.ts:1]: gdi Vlad :|

Comment (vladfrangu) [reply packages/snowflake/src/lib/Snowflake.ts:1]: SORRY 😢

File: packages/snowflake/README.md [hunk 8]
```diff
@@ -9,12 +9,119 @@
...
+#### Snowflake with Discord epoch constant
+
+```ts
+// Import the Snowflake class
+const { DiscordSnowflake } = require('@sapphire/snowflake');
+
+// Create an instance of Snowflake
+const discordSnowflake = new DiscordSnowflake();
+
+// Deconstruct a snowflake with the given epoch
```
Comment (vladfrangu) [inline packages/snowflake/README.md:110]: This should specify it's using Discord's epoch

File: packages/snowflake/README.md [hunk 9]
```diff
@@ -9,12 +9,119 @@
...
+#### Snowflake with Twitter epoch constant
+
+```ts
+// Import the Snowflake class
+const { TwitterSnowflake } = require('@sapphire/snowflake');
+
+// Create an instance of Snowflake
+const twitterSnowflake = new TwitterSnowflake();
+
+// Deconstruct a snowflake with the given epoch
```
Comment (vladfrangu) [inline packages/snowflake/README.md:123]: Ditto

File: packages/snowflake/src/lib/Snowflake.ts [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+	/**
+	 * Alias for {@link deconstruct}
+	 */
+	// eslint-disable-next-line @typescript-eslint/unbound-method, @typescript-eslint/no-invalid-this
+	public decode = this.deconstruct;
+
+	/**
+	 * @param epoch the epoch to use
+	 */
+	public constructor(epoch: bigint) {
```
Comment (vladfrangu) [inline packages/snowflake/src/lib/Snowflake.ts:28]: There's a mismatch between the readme and the types here... You allow people to pass in a Date object, or possibly a number, while taking just a bigint per the types.. can you handle those edge cases? In the end, the epoch should be kept as a bigint internally, but the constructor should handle it

File: packages/snowflake/src/lib/Snowflake.ts [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+		this.#epoch = epoch;
+	}
+
+	/**
+	 * Generates a snowflake given an epoch and optionally a timestamp
+	 * @param __namedParameters options to pass into the generator, see {@link SnowflakeGenerateOptions}
+	 *
+	 * **note** when increment is not provided it defaults to `0n`
```
Comment (vladfrangu) [inline packages/snowflake/src/lib/Snowflake.ts:36]: It defaults to the class' internal increment counter

File: packages/snowflake/src/lib/Snowflake.ts [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+		{ increment = this.#increment, timestamp = Date.now() }: SnowflakeGenerateOptions = { increment: this.#increment, timestamp: Date.now() }
+	) {
+		if (timestamp instanceof Date) timestamp = timestamp.getTime();
+
+		if (typeof timestamp !== 'number' || Number.isNaN(timestamp)) {
+			throw new TypeError(`"timestamp" argument must be a number or Date (received ${Number.isNaN(timestamp) ? 'NaN' : typeof timestamp})`);
+		}
+		if (increment >= 4095n) increment = 0n;
+
+		// timestamp, workerID hard-coded to 1, processID hard-coded to 1, increment
```
Comment (vladfrangu) [inline packages/snowflake/src/lib/Snowflake.ts:54]: As I said, allow us to modify worker ids and process ids via optional options 💙

File: packages/snowflake/src/lib/Snowflake.ts [hunk 13]
```diff
@@ -0,0 +1,130 @@
...
+	 */
+	id: bigint;
+	/**
+	 * The timestamp stored in the snowflake
+	 */
+	timestamp: bigint;
+	/**
+	 * The worker id stored in the snowflake
+	 */
+	workerID: number;
```
Comment (vladfrangu) [inline packages/snowflake/src/lib/Snowflake.ts:114]: Let's make them all bigints for consistency

**Review by vladfrangu — CHANGES_REQUESTED** [review]
You love to see it, mobile review!

File: packages/snowflake/src/lib/Snowflake.ts [hunk 14]
```diff
@@ -0,0 +1,130 @@
...
+	 * ```ts
+	 * const epoch = new Date('2000-01-01T00:00:00.000Z');
+	 * const snowflake = new Snowflake(epoch).deconstruct('3971046231244935168');
+	 * ```
+	 */
+	public deconstruct(id: string | bigint): SnowflakeDeconstruction {
+		const bigIntId = BigInt(id);
+		return {
+			id: bigIntId,
+			timestamp: BigInt(bigIntId >> 22n) + this.#epoch,
```
Comment (vladfrangu) [inline packages/snowflake/src/lib/Snowflake.ts:72]: Also, no need for the BigInt constructor here

**Comment by vladfrangu** [comment]
@all-contributors add @vladfrangu for tests

**Review by kyranet — APPROVED** [review]

**Review by kyranet — DISMISSED** [review]

File: packages/snowflake/src/lib/Snowflake.ts [hunk 15]
```diff
@@ -0,0 +1,151 @@
...
+export interface DeconstructedSnowflake {
+	/**
+	 * The id in BigInt form
+	 */
+	id: bigint;
+	/**
+	 * The timestamp stored in the snowflake
+	 */
+	timestamp: bigint;
+	/**
```
Comment (Favna) [inline packages/snowflake/src/lib/Snowflake.ts:132]:
```suggestion

	/**
```

File: packages/snowflake/src/lib/Snowflake.ts [hunk 16]
```diff
@@ -0,0 +1,151 @@
...
+
+/**
+ * Object returned by Snowflake#deconstruct
+ */
+export interface DeconstructedSnowflake {
+	/**
+	 * The id in BigInt form
+	 */
+	id: bigint;
+	/**
```
Comment (Favna) [inline packages/snowflake/src/lib/Snowflake.ts:128]:
```suggestion

	/**
```

**Review by Favna — COMMENTED** [review]
consistency

**Review by vladfrangu — APPROVED** [review]

**Review by kyranet — APPROVED** [review]

**Pull request merged**